### ISSUE TITLE
Database time records clean up controls. Time. records

### ISSUE THREAD

**Issue by ACIIL** [synkarius/saltbot#126]
If you check you match database see the max and min time values, there is a problem with removing errors.

From my storage i got over 500+ (!) pages of useless  0 time matches, and i got extreme high values high as 348seconds (5min), both of ranges are useless and breaks bot prediction.  

As the bot always weights high towards time, you can see this false time data can poison the bot.  These bad rows also make sim run time longer.

I can use some help here, who is maintaining the database, please create methods to lean up data, or at least point me in the right direction to add my own.

**Comment by reconman** [comment]
The match data is uploaded by random people. 0 means that the time could not be determined. What do you think would be the best behaviour of the saltbot if the time is 0?

You can't use 0 because it would mean that one character instantly won. If you took a maximum value instead you would negatively influence the records with time 0.

**Comment by ACIIL** [comment]
Actually the time data situation gets worse,  in the last few weeks tourneys have been locking up with admins needing to come in and force kills, or admins not appear at all for hours, so those matche's extreme high/low times are also creating false time values.

I think 0 times must be purged but also a high and low time range implemented and ran into the database as cleanup, the lowest would be how fast 3 or 5 forced admin kills are (maybe remove below 20second matches?), highest total time be 3 or 5 round matches lengths .

When such time are detected, purge the match.

**Comment by reconman** [comment]
For the existing records the matches with time 0 would have to be removed from the files manually. For future matches it would be possible to add a logic that records with unrecognizable time are not recorded just like when the winner cannot be determined.

**Comment by ACIIL** [comment]
If i was anymore clever (i am clueless) about to making new menu buttons and using chrome.storage.local, i would make a SQL command that iterates matches finding times outside of a min and max reasonable range, those matches selected will be deleted from the database.

**Comment by bergerst** [comment]
SQL?

Matches can only be overwritten by removing them from the array and setting the matches object in local storage.

**Comment by reconman** [comment]
SQL?

Matches can only be overwritten by removing them from the array and setting the matches object in local storage.

**Comment by ACIIL** [comment]
No naive SQL support?  That is sad to hear.

I was hoping a lead like this resource: https://developers.google.com/web/tools/chrome-devtools/manage-data/local-storage  
meant saltyBot was storing its matches in a IndexedDB or similar when using chrome.storage.  Then i could try my SQL voodoo skills on it.  

I won't stop researching this issue, it is now I question trusting all the input match data for errors as i test ways to improve the heuristic and breeding methods.    I can really use help with this.

**Comment by ACIIL** [comment]
Well i did a purge using some sql tricks in notepad++.  The news is bad.  
Using the new data dump as an example (and modified with some new games...). 

The table set as:
char1,char2,winner,strategy,prediction,tier,mode,odds,time,crowdFavor,illumfavor,date

```
SQL Plugin 1.0.1025
Query         : select * from data where time between 15 and 630
Sourcefile    : sql ready - saltyRecordsM--2017-3-23-0.20.txt
Delimiter     : ,
Number of hits: 26640
```

If my query is correct, about 91100 entries became about 26637. In my own dataset of about 42250 lines became about 11991.

I'm currently make sure the purged list works as import, still debugging stuff.

**Comment by ACIIL** [comment]
Ok i got a guide ready for purging bad times. 

As the data is comma delimited it can treated like a .csv file.  
I ported it in microsoft excel,   and filtered the time column with a between range. Then saved it as a comma delimited .csv, finally renaming it back to .txt. 

This should be a quick fix for cleaning up the current dataset with bad times.

**Comment by reconman** [comment]
Could you create a pull request with the cleaned up data?

**Comment by calexil** [comment]
YES @ACIIL  please pr, this is a disturbing relevation

**Comment by calexil** [comment]
I thought the bot discarded obvious junk match data. ie: 0-12 second matches

**Comment by reconman** [comment]
@calexil I think ACIIL used the SQL function on the wrong column so that's why so many matches were filtered out. Also matches with undetermined length get the time 0 instead of being thrown away.

**Comment by ACIIL** [comment]
Gosh the games data file got huge with that new set at ~5megs, Github is starting to craw as it works with the file. So  I'll  just dump the complete set with my own data as a zip.     (I'm also not sure how to make safe pulls on the main branch: its select branch, commit change, pull request?)

https://www.dropbox.com/s/xveo0c1s8ylrvip/saltyRecordsM--2017-3-23-5.57%28hugelist%20everything%20time%20cleaned%29.txt.zip?dl=0


edit: yup, purge was not as horrifying and removes 10k'ths worth, still, 0 times are useless and are many in numbers. Both as data and browsing the database.

**Comment by calexil** [comment]
do we not have clean chromosome data any more?

**Comment by calexil** [comment]
Ill wrap mine into the zip, I forgot to

**Comment by reconman** [comment]
The SampleData zips still contain chromosomes. I also have some chromosomes on my local filesystem.

**Comment by calexil** [comment]
I have a well evolved set at home, ill merge it into the zip later unless you wanna give something older (keep a little competitive advantage ;) )

**Comment by calexil** [comment]
this can be closed

**Comment by ACIIL** [comment]
Not really, new games are still being inserted at 0 times. The purge never ends.

**Comment by calexil** [comment]
oh duh, you're right, the modes need to be edited to omit 0 time results

probably like 0-5 seconds actually

**Comment by calexil** [comment]
Is your pr gonna fix this or was it just for sorting?

**Comment by ACIIL** [comment]
I am not qualified touching anything related to parsing the saltyBet webpage or the salty Bot manipulation of that data.  I am spending my funs time working on evolution simulation with that i have.    I Discovered that the data was flawed and made this issue so the maintainer can deal with it.

**Comment by ACIIL** [comment]
Think i found a lead for what is happening for time.

```
Match.prototype.update = function (infoFromWaifu, odds, timeInfo, crowdFavor, illumFavor) {
.....

	if (timeInfo.ticks > 0)
		this.time = timeInfo.ticks * timeInfo.interval / 1000;
	//Ignore times from matches that occurred before changing modes; 350 is the maximum time that can occur
	if (this.time >= 240)
		this.time = 0;
	//add more time to matches that are recognized as being in exhibition mode, proportional to the amount of required matches missing
	if (this.mode == "e")
		this.time = Math.round(this.time * 1.5);
.....
```

If this is the cause, Its making unknown matches AND VERY LONG TIMES zero. doh!
Think i can fix this.  I'll just do what i did to the data set. 

I disagree with the max time, i think its 9*70 seconds (full blown potato mode for a final round tourney), but then i need better means to ignore mode changes like the code here attempted.

**Comment by reconman** [comment]
This issue is not fixed, I think. Times are still stored as 0.

**Comment by calexil** [comment]
hmm.. yeah we need to fix that.. the record file gets outta control pretty fast

**Comment by calexil** [comment]
@ACIIL  Can you scrub this file of zero records, I am not sure what you did..

smb://192.168.10.103/saltbot%20downloads/saltyRecordsM--2017-6-19-21.33.txt

**Comment by ACIIL** [comment]
I make a post earlier explaining how to do that, for the purpose of never needing to do it again.   Use the comma delimited .csv file trick.

**Comment by calexil** [comment]
Which field is time?

**Comment by ACIIL** [comment]
From above:
> char1,char2,winner,strategy,prediction,tier,mode,odds,time,crowdFavor,illumfavor,date
Careful in some case, such as treat odds , date, etc as string,

**Comment by calexil** [comment]
the record file is so large I can't even edit it.. :/

**Comment by reconman** [comment]
IndexedDB did nothing to solve the problem, but 0 times are ignored in strategy.ts at the lines 440 to 450. @ACIIL implemented that, I think.
